lib/puppet/util/tagging.rb

```diff
@@ -40,7 +40,7 @@ def tag(*ary)
   # since that results in testing the same string twice
   #
   def tag_if_valid(name)
-    if name.is_a?(String) && name =~ ValidTagRegex
+    if name.is_a?(String) && name =~ ValidTagRegex && Puppet::Resource::MAIN != name
```

Comment:
According to https://docs.puppet.com/puppet/4.10/lang_reserved.html#reserved-class-names, `main` is a reserved class name, though attempts to use it produce a duplicate resource declaration error, which isn't what I was expecting. I assume it's because the implicit `:main` class resolves to `Class[main]{:name=>"main"}`:

```puppet
cat main.pp
class main {
  notify { 'hi': }
}
include main
```

```
$ bx puppet apply main.pp
...
Error: Duplicate declaration: Class[Main] is already declared; cannot redeclare
```